hive-metastore/src/main/java/org/apache/iceberg/hive/HiveLock.java

```diff
@@ -0,0 +1,27 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.hive;
+
+public interface HiveLock {
+  void lock() throws LockException;
+
+  void ensureActive() throws LockException;
+
+  void unlock();
+}
```

Comment:
In #6570 I have described my reasoning behind not using the existing `LockManager` interface:

-----
We already have `LockManager` interface defined in the `iceberg-api` module. After some back-and-forth I decided against using it, because of the following reasons:
- I do not think anyone would like to use HiveLockManager without HiveCatalog
- The interface is not that useful for us:
    - We would need to keep track of the HMS lockId internally
    - We would need to update the LockManager if the `setConf` method is called on the HiveCatalog
    - We would need to add something like `ensureActive` to the interface which is needed for HiveTableOperations
- The BaseLockManager does not provide too much of a functionality
- The current configuration keys are different from the ones used by the LockManager implementations
-----

@szehon-ho and @amogh-jahagirdar, I am happy to change the PR if we can find a good solution. Could you please help me what should be the main approach?
- Shall we keep the `LockManager` as it is and create/use `HiveLockManager` with extra methods, or add `ensureAcitve` method to the `LockManger` interface?
- Shall we keep the current config for the hive locks or shall we use the new ones instead, or keep both configs and create a deprecation handling for it?
- Shall we use the `scheduler` for defined with the `BaseLockManager` and accept the possibility that if the pool is not big enough for the JVM then there will be no heartbeat for the new commits?

Thanks,
Peter

Reply:
@amogh-jahagirdar, @szehon-ho: What do you think? Are any of the options above attractive enough to pursue? Or any other ideas?

Reply:
@pvary sorry for the late reply so here are my thoughts:

> 1.) Shall we keep the LockManager as it is and create/use HiveLockManager with extra methods, or add ensureAcitve method to the LockManger interface?

So after reading the code it seems like LockManager was designed in a manner where the heartbeating is assumed to be happening under the hood for the implementations which require heartbeats. If I'm not mistaken ensureActive is used at certain points to check the heartbeat status. What we could do here is have the separate HiveLockManager interface but it still extends BaseLockManager so we inherit acquire/release, and then the ensureActive is specific to HiveLockManager. We shouldn't have to have a separate lock/unlock imo because acquire/release are already on the LockManager itself, and those are the same operations. 

let me know if that makes sense! 

> 2.) Shall we keep the current config for the hive locks or shall we use the new ones instead, or keep both configs and create a deprecation handling for it?

I'm not super opinionated on this tbh, the current ones make sense to me. If we do decide to add new ones, having a deprecation path then should happen, rather than changing it all at once. 

> 3.) Shall we keep the LockManager as it is and create/use HiveLockManager with extra methods, or add ensureAcitve method to the LockManger interface?

>The current LockManager implementation silently accepts the job, and does not run the heartbeat until there is an empty thread. HiveTableOperations has its own thread and handles this correctly.

This seems like incorrect behavior in the BaseLockManager then, I think we should always guarantee that if the heartbeat thread cannot start we fail. So I think ideally if possible we fix it in the BaseLockmanager itself. In my mind we also add a method `withHeartbeatTask(Runnable heartbeatTask)`.

I think @jackye1995 did the original implementation for BaseLockManager, I'll wait for him to comment though if the silent acceptance of the heartbeat thread is intentional.

Reply:
> I think @jackye1995 did the original implementation for BaseLockManager, I'll wait for him to comment though if the silent acceptance of the heartbeat thread is intentional.

Yeah this is a good point, seems like a miss of the original implementation, we should fix it if it's a known issue

Reply:
I did some drafting myself to see how hard would this be, and here are my thoughts:
- There is a big difference between the `LockManager` which aims to be the single shared instance which is called by multiple `TableOperations` instances (created by the same `Catalog`), and the `MetastoreLock` which is handling a single instance of the lock. With the current code it is possible to have a different locking configuration  for different tables created through the same HiveCatalog implementation (with the use of `HiveCatalog.setConf`), and we aim to keep that, so some tables will use the old locking and some tables will use the new locking mechanism (See: #6570).

For the concrete methods:
- `ensureActive` - this is currently Hive specific, but if you need to have a heartbeat then it means that you run the risk that one way or another the heartbeat fails (temporary connection issue, or whatever). I think this means that failing heartbeat is not Hive specific, just an issue not addressed by other LockManager instances (maybe not that pronounced there but still a valid issue)
- `LockManager`
   - `acquire` HiveLock needs `database` and `table` to acquire a lock - whereas LockManager has only `entityId` - we can get over with it by splitting the `entityId` on `.`, but this is not that nice
   - `release` - we can use this
   - `initialize` - we can get away with pushing the needed stuff through the constructor - but remember that we have to create the LockManager for every operation to ensure that the configurations are honored.
- `BaseLockManager` only implements the following methods:
   - `scheduler` - needs fixing
   - `initialize` - only reads the cofigurations which are not needed anyway for Hive as we plan to keep the old config

So as I see now, the Hive lock handling would not benefit too much from the existing `LockManager` infra.

All-in-all this would be another big change above this big refactor.
If we want to do this we should split it out to multiple steps to help the reviewers.
Also, I am overwhelmed with other work, so I will not have too much time working on this.

@amogh-jahagirdar: I propose to merge this change for now, and if you have time I would be happy to review your PR from the Hive perspective which merges the 2 Locking behaviour.

Would that work for you @amogh-jahagirdar?

Thanks,
Peter

Reply:
@pvary Thanks for the explanation, yeah it should be fine to merge! In this case it does seem complex to reconcile the two. Didn't intend to block there, just wanted to see if we were making the change in the right manner!  We can always perform the refactorings over time as we get more eyes on the classes since these are internal classes.